### PR TITLE
Feature stateless bankster settle overdue acceptances operation

### PR DESCRIPTION
**Pull request by rwrzesien**
Resolves https://github.com/golemfactory/concent/issues/565

### CODE REVIEW

**Comment by Jakub89** [comment]
test

File: concent_api/core/payments/bankster.py [hunk 1]
```diff
@@ -147,3 +157,130 @@ def finalize_payment(
...
+        requestor_eth_address=requestor_ethereum_address,
+        provider_eth_address=provider_ethereum_address,
+        payment_ts=oldest_payments_ts,
+        current_time=current_time,
+        transaction_type=TransactionType.BATCH,
+    )
+
+    # Concent defines time T1 equal to youngest timestamp from list of transactions.
+    if len(list_of_transactions) == 0:
+        t1_is_bigger_than_payments_ts = None
```
Comment (dybi) [inline concent_api/core/payments/bankster.py]: `None` or `False` here?

File: concent_api/core/payments/bankster.py [hunk 2]
```diff
@@ -147,3 +157,130 @@ def finalize_payment(
...
+        )
+
+    # Any of the items from list of overdue acceptances
+    # matches condition current_time < payment_ts + PAYMENT_DUE_TIME
+    acceptance_time_overdue = any(
+        current_time < subtask_results_accepted.payment_ts + settings.PAYMENT_DUE_TIME
+        for subtask_results_accepted in acceptances
+    )
+
+    if t1_is_bigger_than_payments_ts or acceptance_time_overdue:
```
Comment (dybi) [inline concent_api/core/payments/bankster.py]: as we are using variables in logical context, both of them should be of type: `bool`. `t1_is_bigger_than_payments_ts` can be `None`

File: concent_api/core/payments/bankster.py [hunk 3]
```diff
@@ -147,3 +157,130 @@ def finalize_payment(
...
+        list_of_payments=list_of_transactions,
+        subtask_results_accepted_list=acceptances,
+    )
+
+    if amount_pending == 0:
+        requestors_claim_payment_info = ClaimPaymentInfo(0, 0)
+    elif requestor_deposit_value == 0:
+        requestors_claim_payment_info = ClaimPaymentInfo(0, amount_pending)
+    else:
+        request_can_pay = min(amount_pending, requestor_deposit_value)
```
Comment (dybi) [inline concent_api/core/payments/bankster.py]: `requestor_payable_amount`? It's name right now suggest that it's `bool` whereas it is in fact an `int`.

File: concent_api/core/payments/bankster.py [hunk 4]
```diff
@@ -147,3 +157,130 @@ def finalize_payment(
...
+    return sum([item.amount for item in payments])
+
+
+def sum_subtask_price(subtask_results_accepted_list: List[SubtaskResultsAccepted]) -> int:
+    assert isinstance(subtask_results_accepted_list, list)
+
+    return sum([subtask_results_accepted.task_to_compute.price for subtask_results_accepted in subtask_results_accepted_list])
+
+
+def sum_amount_price_for_provider(
```
Comment (dybi) [inline concent_api/core/payments/bankster.py]: mhm. maybe `get_provider_payment_info`? or `get_info_about_payment_for_provider`?

File: concent_api/core/tests/test_payments_bankster.py [hunk 5]
```diff
@@ -269,3 +258,108 @@ def test_that_in_not_additional_verification_use_case_provider_claim_payment_inf
...
+                        acceptances=subtask_results_accepted_list,
+                    )
+
+        get_deposit_value_mock.assert_called_once()
+        get_list_of_payments_mock.assert_called()
+
+        self.assertIsNotNone(requestors_claim_payment_info.tx_hash)
+        self.assertIsNotNone(requestors_claim_payment_info.payment_ts)
+        self.assertEqual(requestors_claim_payment_info.amount_paid, 1000)
+        self.assertEqual(requestors_claim_payment_info.amount_pending, 1000)
```
Comment (dybi) [inline concent_api/core/tests/test_payments_bankster.py]: shouldn't `amount_pending` be equal to 1500 (price) - 1000 (amount_paid) = 14000?

File: concent_api/core/payments/bankster.py (hunk 1 above)
Comment (rwrzesien) [reply concent_api/core/payments/bankster.py]: :+1: done

File: concent_api/core/payments/bankster.py (hunk 2 above)
Comment (rwrzesien) [reply concent_api/core/payments/bankster.py]: :+1: done

File: concent_api/core/payments/bankster.py (hunk 3 above)
Comment (rwrzesien) [reply concent_api/core/payments/bankster.py]: :+1: done

File: concent_api/core/payments/bankster.py (hunk 4 above)
Comment (rwrzesien) [reply concent_api/core/payments/bankster.py]: :+1: done

File: concent_api/core/tests/test_payments_bankster.py (hunk 5 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_payments_bankster.py]: No, please check `_get_list_of_batch_transactions` and `_get_list_of_force_transactions` mock functions in this tests. Calculations are like this:

```
already_paid_value = 15000 - (
    (1000 + 2000 + 3000 + 4000)(batch_transactions) +
    (1000 + 2000)(force_transactions) +
)
already_paid_value == 13000, so 2000 left
get_deposit_value returns 1000, so 1000 paid and 1000 left (pending)
```

File: concent_api/core/payments/backends/mock.py [hunk 6]
```diff
@@ -20,8 +20,18 @@ def make_force_payment_to_provider(
     provider_eth_address: str,  # pylint: disable=unused-argument
     value: int,  # pylint: disable=unused-argument
     payment_ts: int,  # pylint: disable=unused-argument
-) -> None:
-    pass
+) -> Transaction:
+    return Transaction(
```
Comment (dybi) [inline concent_api/core/payments/backends/mock.py]: BTW: maybe `MOCK_TRANSACTION` can be reused here?

File: concent_api/core/tests/test_payments_bankster.py (hunk 5 above)
Comment (dybi) [reply concent_api/core/tests/test_payments_bankster.py]: can you please put in in a comment (in a short from)? It is not visible at a first glance

File: concent_api/core/payments/backends/mock.py (hunk 6 above)
Comment (rwrzesien) [reply concent_api/core/payments/backends/mock.py]: :+1: done

File: concent_api/core/tests/test_payments_bankster.py (hunk 5 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_payments_bankster.py]: :+1: done

File: concent_api/core/payments/bankster.py [hunk 7]
```diff
@@ -147,3 +157,130 @@ def finalize_payment(
...
+def sum_subtask_price(subtask_results_accepted_list: List[SubtaskResultsAccepted]) -> int:
+    assert isinstance(subtask_results_accepted_list, list)
+
+    return sum([subtask_results_accepted.task_to_compute.price for subtask_results_accepted in subtask_results_accepted_list])
+
+
+def get_provider_payment_info(
+    list_of_forced_payments:        List[ForcedPaymentEvent],
+    list_of_payments:               List[BatchTransferEvent],
+    subtask_results_accepted_list:  List[SubtaskResultsAccepted],
```
Comment (Jakub89) [inline concent_api/core/payments/bankster.py]: No need for old style :)

File: concent_api/core/tests/test_auth_force_payment.py [hunk 8]
```diff
@@ -58,26 +59,17 @@ def test_provider_send_force_payment_and_concent_should_return_it_to_requestor_w
...
-                        data                                = serialized_force_payment,
-                        content_type                        = 'application/octet-stream',
-                    )
-
-        make_force_payment_to_provider_mock.assert_called_once()
-        self.assertEqual(get_list_of_payments_mock.call_count, 2)
+                response_1 = self.client.post(
+                    reverse('core:send'),
+                    data                                = serialized_force_payment,
+                    content_type                        = 'application/octet-stream',
```
Comment (Jakub89) [inline concent_api/core/tests/test_auth_force_payment.py]: whitespace

Comment (rwrzesien) [reply concent_api/core/tests/test_auth_force_payment.py]: :+1: done

File: concent_api/core/payments/bankster.py (hunk 7 above)
Comment (rwrzesien) [reply concent_api/core/payments/bankster.py]: :+1: done

**Pull request merged**